MessageLogger/src/main/java/io/github/juby210/acplugins/messagelogger/SQLite.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (c) 2021-2022 Juby210
+ * Licensed under the Open Software License version 3.0
+ */
+
+package io.github.juby210.acplugins.messagelogger;
+
+import android.content.Context;
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+import android.database.sqlite.SQLiteOpenHelper;
+import android.os.Environment;
+
+import com.aliucord.Utils;
+import com.aliucord.wrappers.ChannelWrapper;
+import com.discord.api.channel.Channel;
+import com.discord.models.deserialization.gson.InboundGatewayGsonParser;
+
+import java.io.File;
+import java.io.IOException;
+
+public final class SQLite extends SQLiteOpenHelper {
+
+    private static final String DB_NAME = "message_logger.db";
+
+    private static final int DB_VERSION = 1;
+
+    private static final String TABLE_NAME = "deleted_messages";
+    private static final String TABLE_NAME_EDITS = "edited_messages";
+    private static final String TABLE_NAME_GUILDS = "guilds";
+    private static final String TABLE_NAME_CHANNELS = "channels";
+    private static final String TABLE_NAME_SETTINGS = "settings";
+
+    public SQLite(Context context) {
+        super(context, DB_NAME, null, DB_VERSION);
+    }
+
+    @Override
+    public void onCreate(SQLiteDatabase db) {
+        String query = "CREATE TABLE " + TABLE_NAME + " (" +
+            "id LONG, " +
+            "delete_data, " +
+            "record TEXT)";
+        db.execSQL(query);
+        query = "CREATE TABLE " + TABLE_NAME_EDITS + " (" +
+            "id LONG, " +
+            "record TEXT)";
+        db.execSQL(query);
+        query = "CREATE TABLE " + TABLE_NAME_GUILDS + " (" +
+            "id LONG)";
+        db.execSQL(query);
+        query = "CREATE TABLE " + TABLE_NAME_CHANNELS + " (" +
+            "id LONG)";
+        db.execSQL(query);
+        query = "CREATE TABLE " + TABLE_NAME_SETTINGS + " (" +
+            "name TEXT, " +
+            "value TEXT)";
+        db.execSQL(query);
+    }
+
+    public void addNewMessage(MessageRecord record) {
+        SQLiteDatabase db = getWritableDatabase();
+        String recordJson = InboundGatewayGsonParser.toJson(record);
+        String deleteDataJson = InboundGatewayGsonParser.toJson(record.deleteData);
+        String query = "INSERT INTO " + TABLE_NAME + " (id, delete_data, record) VALUES (?, ?, ?)";
+        db.execSQL(query, new Object[]{record.message.getId(), deleteDataJson, recordJson});
+    }
+
+    public void addNewMessageEdit(MessageRecord record) {
+        SQLiteDatabase db = getWritableDatabase();
+        String recordJson = InboundGatewayGsonParser.toJson(record);
+        String query = "SELECT * FROM " + TABLE_NAME_EDITS + " WHERE id = ?";
+        Cursor cursor = db.rawQuery(query, new String[]{String.valueOf(record.message.getId())});
+        if (cursor.getCount() > 0) {
```

Comment:
this is also bad style because updates aren't atomic. It could happen that between the select and update, another thread inserts. You should do a proper upsert https://www.sqlite.org/lang_UPSERT.html

Reply:
looks like upset requires API 30 or higher

in that case, first try to yolo the update and if no rows were affected do an insert

Reply:
> this is also bad style because updates aren't atomic. It could happen that between the select and update, another thread inserts. You should do a proper upsert https://www.sqlite.org/lang_UPSERT.html

ill look into changing this when i have some free time

Reply:
Fixed in new pr